### ISSUE TITLE
Investigate parsing strategy in MsgPack related to ReadOnlyCollections

### ISSUE THREAD

**Issue by BrennanConroy** [aspnet/SignalR#862]
I had a streaming hub method returning `IReadOnlyCollection<T>` when using MessagePack a parsing exception occurs. We need to investigate this and figure out how to proceed.

```
System.Runtime.Serialization.SerializationException: Type 'System.Collections.ObjectModel.ReadOnlyCollection`1[Microsoft.AspNet.SignalR.StockTicker.Stock]' does
 not have both of default (parameterless) public constructor and  public constructor with an Int32 parameter.
   at MsgPack.Serialization.SerializationExceptions.ThrowTargetDoesNotHavePublicDefaultConstructorNorInitialCapacity(Type type)
   at MsgPack.Serialization.UnpackHelpers.DetermineCollectionSerializationStrategy(Type instanceType, Boolean allowAsymmetricSerializer)
   at MsgPack.Serialization.ReflectionSerializers.ReflectionSerializerHelper.CreateCollectionSerializer[T](SerializationContext context, Type targetType, CollectionTraits traits, PolymorphismSchema schema)
   at MsgPack.Serialization.MessagePackSerializer.CreateReflectionInternal[T](SerializationContext context, Type concreteType, PolymorphismSchema schema)
   at MsgPack.Serialization.MessagePackSerializer.CreateInternal[T](SerializationContext context, PolymorphismSchema schema)
   at MsgPack.Serialization.SerializationContext.GetSerializer[T](Object providerParameter)
   at MsgPack.Serialization.SerializationContext.SerializerGetter.Get(SerializationContext context, Type targetType, Object providerParameter)
   at MsgPack.PackerUnpackerExtensions.PackObject(Packer source, Object value)
   at Microsoft.AspNetCore.SignalR.Internal.Protocol.MessagePackHubProtocol.WriteStreamingItemMessage(StreamItemMessage streamItemMessage, Packer packer, Stream output) in C:\Github\SignalR\src\Microsoft.AspNetCore.SignalR.Common\Internal\Protocol\MessagePackHubProtocol.cs:line 164
   at Microsoft.AspNetCore.SignalR.Internal.Protocol.MessagePackHubProtocol.WriteMessageCore(HubMessage message, Stream output) in C:\Github\SignalR\src\Microsoft.AspNetCore.SignalR.Common\Internal\Protocol\MessagePackHubProtocol.cs:line 139
   at Microsoft.AspNetCore.SignalR.Internal.Protocol.MessagePackHubProtocol.WriteMessage(HubMessage message, Stream output) in C:\Github\SignalR\src\Microsoft.AspNetCore.SignalR.Common\Internal\Protocol\MessagePackHubProtocol.cs:line 125
   at Microsoft.AspNetCore.SignalR.Internal.HubProtocolReaderWriter.WriteMessage(HubMessage hubMessage) in C:\Github\SignalR\src\Microsoft.AspNetCore.SignalR.Common\Internal\HubProtocolReaderWriter.cs:line 32
   at Microsoft.AspNetCore.SignalR.HubEndPoint`1.<>c__DisplayClass11_0.<<OnConnectedAsync>g__WriteToTransport0>d.MoveNext() in C:\Github\SignalR\src\Microsoft.AspNetCore.SignalR.Core\HubEndPoint.cs:line 83
```

**Comment by BrennanConroy** [comment]
In `packer.PackObject(...)` we can pass a `SerializationContext` and set `SerializationContext.CompatibilityOptions.AllowAsymmetricSerializer = true` which allows more types through the serializer checker

**Comment by moozzyk** [comment]
Fixed in dbe4b4025bb9518f816f4762bf53c1f8ae02407e
